weave/trace_server/errors.py

```diff
@@ -39,15 +76,227 @@ def __init__(self, message: str, api_key_name: str):
         super().__init__(message)
 
 
-class NotFoundError(Error):
-    """Raised when a general not found error occurs."""
-
-    pass
-
-
 class ObjectDeletedError(Error):
     """Raised when an object has been deleted."""
 
     def __init__(self, message: str, deleted_at: datetime.datetime):
         self.deleted_at = deleted_at
         super().__init__(message)
+
+
+class InvalidExternalRef(Error):
+    """Raised when an external reference is invalid."""
+
+    pass
+
+
+class ProjectNotFound(Error):
+    """Raised when a project is not found."""
+
+    pass
+
+
+def _format_error_to_json(exc: Exception) -> dict[str, Any]:
+    """Default formatter that tries to parse JSON or falls back to reason."""
+    exc_str = str(exc)
+    try:
+        return json.loads(exc_str)
+    except json.JSONDecodeError:
+        return {"reason": exc_str}
+
+
+def _format_error_to_json_with_extra(
+    exc: Exception, extra_fields: dict[str, Any]
+) -> dict[str, Any]:
+    """Helper to format exception as JSON or fallback to reason, always adding extra fields."""
+    result = _format_error_to_json(exc)
+    result.update(extra_fields)
+    return result
```

Comment:
Merge those two into a single function that accepts an optional `extra_fields` arg.